### PR TITLE
EM-112: merge Inventory DTOs when suID and partNo are the same

### PR DESCRIPTION
**Pull request by jhoule86**
Bug fixes for "Multiple warehouses in inventory" (https://github.com/wayfair-contribs/plentymarkets-plugin/issues/51).
As pointed out by Anish, the supplier may purposely enter the same Supplier ID to 2 or more warehouses. We should report aggregate data for these Warehouses in a single DTO, not in separate DTOs per Warehouse.

**Also resurrects support for the `Stock Buffer` setting!** (see https://github.com/wayfair-contribs/plentymarkets-plugin/issues/50)

### CODE REVIEW

File: src/Mappers/InventoryMapper.php [hunk 1]
```diff
@@ -110,19 +111,50 @@ public function createInventoryDTOsFromVariation($variationData)
...
+      {
+        // merge with previous values for this suID
+        if (isset($existingDTO->getQuantityOnHand()))
+        {
+          $onHand += $existingDTO->getQuantityOnHand();
+        }
+
+        if (isset($existingDTO->getQuantityOnOrder()))
+        {
+          $onOrder += $existingDTO;
```
Comment (mcloughlin77) [inline src/Mappers/InventoryMapper.php:135]: This should be $onOrder += $existingDTO->getQuantityOnOrder();

File: src/Mappers/InventoryMapper.php [hunk 2]
```diff
@@ -110,19 +111,50 @@ public function createInventoryDTOsFromVariation($variationData)
...
+        }
+
+        if (!isset($nextAvailableDate))
+        {
+          $nextAvailableDate = $existingDTO->getItemNextAvailabilityDate();
```
Comment (mcloughlin77) [inline src/Mappers/InventoryMapper.php:140]: For next available date should we take whichever date is earliest?

File: src/Mappers/InventoryMapper.php (hunk 1 above)
Comment (jhoule86) [reply src/Mappers/InventoryMapper.php:135]: oh. good call.

File: src/Mappers/InventoryMapper.php (hunk 2 above)
Comment (jhoule86) [reply src/Mappers/InventoryMapper.php:140]: I originally felt that way as well, but the next available date is coming from the variation level (outside of this loop), not the individual warehouse.

File: src/Mappers/InventoryMapper.php [hunk 3]
```diff
@@ -71,7 +72,7 @@ public function createInventoryDTOsFromVariation($variationData)
     /** @var LoggerContract $loggerContract */
     $loggerContract = pluginApp(LoggerContract::class);
 
-    /** @var RequestDTO[] $inventoryDTOs */
+    /** @var array $inventoryDTOs */
```
Comment (dFed17) [inline src/Mappers/InventoryMapper.php:75]: Is RequestDTO[] a sniff currently? Not an issue - just curious

File: src/Mappers/InventoryMapper.php (hunk 2 above)
Comment (jhoule86) [reply src/Mappers/InventoryMapper.php:140]: so actually yeah this code doesn't do anything in the first place. whoops.

File: src/Mappers/InventoryMapper.php (hunk 3 above)
Comment (dFed17) [reply src/Mappers/InventoryMapper.php:75]: Also would consider calling this variable `$indexedRequestDTOs` or something - the `array_value` call in the return was a little confusing for me cause this wasn't clear

Comment (jhoule86) [reply src/Mappers/InventoryMapper.php:75]: AFAIK `RequestDTO[]` is not a vaild type hint for associative arrays. I could be wrong.
Yeah I agree about the name. Something like `requestDtosBySuID` is more descriptive.

Comment (jhoule86) [reply src/Mappers/InventoryMapper.php]: found out that I can do `@var array<string,RequestDTO> $requestDtosBySuID`

File: src/Mappers/InventoryMapper.php [hunk 4]
```diff
@@ -110,19 +111,48 @@ public function createInventoryDTOsFromVariation($variationData)
...
+      if (isset($existingDTO) && !empty($existingDTO))
+      {
+        /* merge with previous values for this suID.
+         * stock values should be summed.
+         * Variation-level data (nextAvailableDate) does not vary.
+        */
+        if (null !== $existingDTO->getQuantityOnHand())
+        {
+          $onHand += $existingDTO->getQuantityOnHand();
+        }
```
Comment (mcloughlin77) [inline src/Mappers/InventoryMapper.php:134]: As -1 can be an option we will need to handle this scenario. My initial thoughts are:

- If both are -1 then quantity is -1
- If one is -1 and the other is 0 then quantity is -1
- If one is -1 and the other is greater than 0 then quantity is the positive quantity

Comment (jhoule86) [reply src/Mappers/InventoryMapper.php]: @mcloughlin77 check out the new static function and associated test

File: src/Mappers/InventoryMapper.php [hunk 5]
```diff
@@ -42,26 +43,63 @@ public function getAvailableDate($mainVariationId)
...
-   * @param mixed $variationStock
+   * Determine the "Quantity On Hand" to report to Wayfair,
+   * based on a VariationStock
+   * @param VariationStock $variationStock
+   * @param AbstractConfigHelper $configHelper
    *
    * @return int|mixed
    */
-  public function getNetStock($variationStock)
+  static function getQuantityOnHand($variationStock, $configHelper)
```
Comment (jhoule86) [inline src/Mappers/InventoryMapper.php:53]: make configHelper optional here

File: src/Mappers/InventoryMapper.php [hunk 6]
```diff
@@ -110,19 +152,53 @@ public function createInventoryDTOsFromVariation($variationData)
         continue;
       }
 
+      // Avl Immediately. Net Stock.
```
Comment (jhoule86) [inline src/Mappers/InventoryMapper.php:155]: update comment to say adjusted net stock

File: src/Mappers/InventoryMapper.php [hunk 7]
```diff
@@ -169,4 +245,44 @@ private function getSupplierPartNumberFromVariation($variationData)
...
+   * @param [bool] $nullable
+   * @return float|null
+   */
+  static function mergeInventoryQuantities($left, $right)
+  {
+    // protecting against values below -1
+    if (null != $left && $left < -1)
+    {
+      // TODO: add log?
+      $left = 0;
```
Comment (jhoule86) [inline src/Mappers/InventoryMapper.php:266]: things below -1 should default to -1

File: src/Services/InventoryUpdateService.php [hunk 8]
```diff
@@ -69,10 +69,18 @@ private function validateInventoryRequestData($inventoryRequestDTO, $loggerContr
...
+    }
+
     if (!isset($issues) || empty($issues)) {
       return true;
     }
 
+    // TODO: translate the issues?
     $loggerContract
       ->error(
         TranslationHelper::getLoggerKey(self::LOG_KEY_INVENTORY_UPDATE_ERROR),
```
Comment (jhoule86) [inline src/Services/InventoryUpdateService.php:86]: make sure this is the right log key for invalid

File: tests/php/Mappers/InventoryMapperTest.php [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+            [0, -1, -1, "zero and negative 1 should make negative 1"],
+            [-1, 1, 1, "negative 1 and 1 should make positive 1"],
+            [1, -1, 1, "1 and negative 1 should make postiive 1"],
+            [0, 1, 1, "zero and 1 should make 1"],
+            [1, 0, 1, "1 and zero should make 1"],
+            [1, 1, 2, "1 and 1 should make 2"],
+            [5, 9, 14, "5 and 9 should make 14"],
+            [9, 5, 14, "9 and 5 should make 14"],
+            [-1, -1, -1, "two negative 1s should make -1"],
+            [-2, -2, 0, "two negative 2s should make 0"],
```
Comment (jhoule86) [inline tests/php/Mappers/InventoryMapperTest.php:46]: should make -1 !

File: src/Mappers/InventoryMapper.php [hunk 10]
```diff
@@ -42,26 +43,63 @@ public function getAvailableDate($mainVariationId)
...
+    {
+      // no buffer to apply
+      return $netStock;
+    }
+
+
+    if ($stockBuffer < 0)
+    {
+      // invalid value for buffer
+      // TODO: add warning log
```
Comment (jhoule86) [inline src/Mappers/InventoryMapper.php:86]: do this

File: src/Mappers/InventoryMapper.php [hunk 11]
```diff
@@ -110,19 +152,53 @@ public function createInventoryDTOsFromVariation($variationData)
         continue;
       }
 
+      // Avl Immediately. Net Stock.
+      $onHand = $this->getQuantityOnHand($stock, $configHelper);
+
+      if (null == $onHand)
+      {
+        // null value is NOT a valid input for quantity on hand - do NOT send to Wayfair.
+        // TODO: add log?
```
Comment (jhoule86) [inline src/Mappers/InventoryMapper.php:161]: do this

File: src/Services/InventoryUpdateService.php [hunk 12]
```diff
@@ -69,13 +55,42 @@ private function validateInventoryRequestData($inventoryRequestDTO, $loggerContr
...
+    else
+    {
+      $issues[] = "Quantity On Hand is missing";
+    }
+
     if (!isset($issues) || empty($issues)) {
       return true;
     }
 
+    // TODO: replace issues with translated messsages?
```
Comment (mcloughlin77) [inline src/Services/InventoryUpdateService.php:90]: Are we waiting until after the logging size issue is fixed to do this?

**Pull request merged**